network/address.go

```diff
@@ -67,7 +68,81 @@ func (a Address) NetworkAddress() string {
 		return ""
 	}
 	vals := strings.Split(string(a), typeAddressSep)
-	return vals[1]
+	validNetworkAddress := vals[1]
+
+	host, port, err := net.SplitHostPort(validNetworkAddress)
+	if err != nil {
+		return ""
+	}
+
+	// checking the validity of the IP address (both IPv4 or IPv6)
+	if net.ParseIP(host) != nil {
+		return validNetworkAddress
+	}
+
+	// if host is not a valid IP address (i.e. ParseIP returns nil),
+	// it must be a hostname (since at the beginning we check whether the address is valid
+	ipAddress, err := net.LookupHost(host)
+	if err != nil {
+		return ""
+	}
+
+	sort.Strings(ipAddress)
+	// it returns the first found ip address
+	return net.JoinHostPort(ipAddress[0], port)
+}
+
+// validHostname returns true if the hostname is well formed or false otherwise.
+// A hostname is well formed if the following conditions are met:
+//	- each label contains from 1 to 63 characters
+//	- the entire hostname (including the delimiting dots, but not a trailing dot)
+// 		has a maximum of 253 ASCII characters
+//	- labels have only ASCII letters 'a' through 'z' (case-insensitive), the digits
+//		'0' through '9', and the hyphen (-). No other symbol is permitted
+//	- labels cannot start with a hyphen
+//	- labels cannot end with a hyphen
+//	- the last label is alphabetic
+// sources about the syntax of hostnames:
+//	- https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names
+//	- https://en.wikipedia.org/wiki/List_of_Internet_top-level_domains#IDN_Test_TLDs
+//	- RFC 1035, RFC 1123
+//	Theoretically, it seems that a TLD with a hyphen (-) is accepted:
```

Comment:
All of the below discussion can be removed from the code. Specify as much as possible in the issue related to this PR and add a link to the issue here so we can reference it later.